smithy-diff/src/main/java/software/amazon/smithy/diff/evaluators/ChangedNullability.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.diff.evaluators;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.StringJoiner;
+import java.util.stream.Stream;
+import software.amazon.smithy.diff.ChangedShape;
+import software.amazon.smithy.diff.Differences;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.NullableIndex;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.traits.DefaultTrait;
+import software.amazon.smithy.model.traits.InputTrait;
+import software.amazon.smithy.model.traits.NullableTrait;
+import software.amazon.smithy.model.traits.RequiredTrait;
+import software.amazon.smithy.model.validation.ValidationEvent;
+
+/**
+ * Validates that only backward compatible changes are made to
+ * structure member nullability to ensure that if something was
+ * previously nullable to clients then it continue to be nullable
+ * and vice versa.
+ */
+public class ChangedNullability extends AbstractDiffEvaluator {
+    @Override
+    public List<ValidationEvent> evaluate(Differences differences) {
+        NullableIndex oldIndex = NullableIndex.of(differences.getOldModel());
+        NullableIndex newIndex = NullableIndex.of(differences.getNewModel());
+
+        List<ValidationEvent> events = new ArrayList<>();
+
+        Stream<ChangedShape<MemberShape>> changed = Stream.concat(
+            // Get members that changed.
+            differences.changedShapes(MemberShape.class),
+            // Get members of structures that added/removed the input trait.
+            changedInputMembers(differences)
```

Comment:
Do members not count as having changed if their traits change?

Reply:
Yeah, any change to a member gets picked up